doc/sphinx/language/core/modules.rst

```diff
@@ -511,6 +511,258 @@ transparent constraint
+      only through their fully qualified names rather than their
+      unqualified names after an :cmd:`Import`.
+
+.. attr:: export
+
+   This :term:`attribute` makes the effect of the command
+   persist when the section is closed and applies the effect when the
+   module containing the command is imported.
+
+   Commands supporting this attribute include :cmd:`Set`, :cmd:`Unset`
+   and the :ref:`Hint <creating_hints>` commands, although the latter
+   don't support it within sections.
+
+.. attr:: global
+
+   This :term:`attribute` makes the effect of the command
+   persist even when the current section or module is closed.  Loading
+   the file containing the command (possibly transitively) applies the
+   effect of the command.
+
+   The ``Global`` prefix is an alternative syntax for the
+   :attr:`global` attribute (see :n:`@legacy_attr`).
+
+   .. warning::
+
+      **Exception:** for a few commands (like :cmd:`Notation` and
+      :cmd:`Ltac`), this attribute behaves like :attr:`export`.
+
+   .. warning::
+
+      We strongly discourage using the :attr:`global` locality
+      attribute because the transitive nature of file loading gives
+      the user little control. We recommend using the :attr:`export`
+      locality attribute where it is supported.
+
+.. _visibility-attributes-modules:
+
+Summary of locality attributes in a module
+------------------------------------------
+
+This table sums up the effect of locality attributes on the scope of vernacular
+commands in a module, when outside the module where they were entered. A cross
+(❌) marks an unsupported attribute, which will provoke a compilation error. In
+this table, "not available", means that the command has no effect outside the
+module it was entered and "short name when imported" means that the command
+always has effects outside the module but if the module (or the command, via
+selective importation, when available) is not imported, the corresponding
+identifier must be qualified in order to be used.
+
+A similar table for :cmd:`Section` can be found
+:ref:`here<visibility-attributes-sections>`.
+
+.. list-table::
+  :header-rows: 1
+
+  * - ``Command``
+    - no attribute
+    - :attr:`local`
+    - :attr:`export`
+    - :attr:`global`
+
+  * - :cmd:`Definition`, :cmd:`Lemma`, :cmd:`Axiom`, ...
+    - :attr:`global`
+    - qualified name
+    - ❌
+    - short name when imported
+
+  * - :cmd:`Notation`
+    - :attr:`global`
+    - not available
+    - ❌
+    - when imported
+
+  * - :cmd:`Notation (abbreviation)`
+    - :attr:`global`
+    - not available
+    - ❌
+    - short name when imported
+
+  * - ``Hints`` (and :cmd:`Instance`)
+    - :attr:`export`
+    - not available
+    - when imported
+    - always
+
+  * - :cmd:`Set` or :cmd:`Unset` a flag
+    - :attr:`local`
+    - not available
+    - when imported
+    - always
+
+  * - :cmd:`Canonical Structure`
+    - :attr:`local`
+
+      or :attr:`global`
```

Comment:
For this one, I just saw experimentally that default, local and global have the same meaning for the Canonical command. See also issue #19348. I did not manage to find the .ml file where this default locality attribute is fixed. @SkySkimmer would you know where to find this piece of information?

Reply:
https://github.com/coq/coq/blob/e564bef2404475ec83061cf7bf4b0759d58aca85/vernac/canonical.ml#L30

`cache` is run when adding the object and at the end of sections if it's kept
`open` is run when imported, in this case it checks `Int.equal i 1` which means it does nothing for deep imports (eg if the object is from M.N when importing M)
`load` is run when requiring and at the end of modules, in this case it's not provided so it does nothing
`subst` is module substitution, doesn't matter for locality
`classify` tells if the object is kept at the end of modules, in this case it's Substitute so it's always kept. Dispose would mean it's dropped. There's also Keep which means it's kept but doesn't work with functors, and Anticipate which is only used for Require because it needs to coordinate with the kernel
`discharge` tells if the object is kept at the end of sections, in this case it depends on the locality boolean

So for the canonical structure object we see
- `local` is default false (unless the callers have some other default, in this case they don't)
- it is kept at the end of sections if not local and not a variable from the section, so eg `Section S. Variable x : ... Canonical Structure x. End S. (* not kept *)`
- it is always kept at the end of modules and takes effect on non-deep Import, not Require

See also https://github.com/coq/coq/blob/e564bef2404475ec83061cf7bf4b0759d58aca85/library/libobject.mli#L15-L65

Reply:
Thank you very much for this very informative answer!

I'll settle for "local", then, it's consistent with what is in sections and closer to what happens (the Canonical is available outside the module by default).